### PR TITLE
Deprecate the final_step_result parameter of TrialResult

### LINKED ISSUE #5271: Deprecate parameter final_step_result from TrialResult.init
**Summarize the task**

Follow-up to #5260, as that PR makes final_sim_state unnecessary. Quite a few tests use this so doing this in a separate PR to keep the scope of the original PR smaller.

Also deprecate DensityMatrixSimulatorState and StateVectorSimulatorState.

### PR DESCRIPTION
**Pull request by daxfohl**
Follow-up to #5260, as that PR makes final_step_result unnecessary.

Also deprecate DensityMatrixSimulatorState and StateVectorSimulatorState

Closes #5271

### CODE REVIEW

**Comment by daxfohl** [comment]
@95-martin-orion This ended up being bigger than I thought, but should end up making things simpler once all these things get fully removed. The removal of mocks from a bunch of tests is a good sign. No rush on this one.

File: cirq-core/cirq/sim/simulator.py [hunk 1]
```diff
@@ -876,6 +889,55 @@ def sample_measurement_ops(
         )
 
 
+# When removing this, also remove the check in simulate_sweep_iter.
+# Basically there should be no "final_step_result" anywhere in the project afterwards.
+def _deprecated_step_result_parameter(
```
Comment (95-martin-orion) [inline cirq-core/cirq/sim/simulator.py:894]: The intent is to remove this method after the next release, right?

**Review by 95-martin-orion — APPROVED** [review]
One question, just for my own benefit. Rename of ActOnArgs to follow?

File: cirq-core/cirq/sim/simulator.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/simulator.py:894]: Yes, this was purely to reduce the redundancy of those deprecation attributes. It's not intended to be general purpose or used for anything else ever, and will go away when the deprecated parameters do.

Comment (daxfohl) [reply cirq-core/cirq/sim/simulator.py:894]: After this, #5270 will be a small PR, then ActOnArgs rename after that.

**Pull request merged**